File: 문제풀이/2025/030225 백준 28472 Minimax Tree G3.py
from sys import setrecursionlimit, stdin

setrecursionlimit(10**6)


def max_val(tree, node, values, computed):
    values[node] = max(
        map(
            lambda node: (
                values[node]
                if computed[node]
                else min_val(tree, node, values, computed)
            ),
            tree[node],
        )
    )

    computed[node] = True
    return values[node]


def min_val(tree, node, values, computed):
    values[node] = min(
        map(
            lambda node: (
                values[node]
                if computed[node]
                else max_val(tree, node, values, computed)
            ),
            tree[node],
        )
    )

    computed[node] = True
    return values[node]


def main():
    data = stdin.read().strip().splitlines()
    nodes, root = map(int, data[0].split())

    graph = [[] for _ in range(nodes + 1)]
    for i in range(1, nodes):
        u, v = map(int, data[i].split())
        graph[u].append(v)
        graph[v].append(u)

    tree = [[] for _ in range(nodes + 1)]
    stack = [root]
    visited = [False] * (nodes + 1)
    visited[root] = True

    while stack:
        node = stack.pop()
        for nxt in graph[node]:
            if not visited[nxt]:
                visited[nxt] = True
                tree[node].append(nxt)
                stack.append(nxt)

    index = nodes
    l = int(data[index].strip())
    index += 1

    values = [0] * (nodes + 1)
    computed = [False] * (nodes + 1)

    for _ in range(l):
        node, val = map(int, data[index].split())
        index += 1
        values[node] = val
        computed[node] = True

    max_val(tree, root, values, computed)

    q = int(data[index].strip())
    index += 1
    results = []

    for _ in range(q):
        node = int(data[index].split()[0])
        index += 1
        results.append(str(values[node]))

    print("\n".join(results))


if __name__ == "__main__":
    main()

File: 문제풀이/2025/030225 백준 28472 Minimax Tree G3.rs
fn parse<T1: std::str::FromStr, T2: std::iter::FromIterator<T1>>(line: &str) -> T2
where
    <T1 as std::str::FromStr>::Err: std::fmt::Debug,
{
    line.trim()
        .split_ascii_whitespace()
        .map(|word| T1::from_str(word).unwrap())
        .collect()
}

fn input<T1: std::str::FromStr, T2: std::iter::FromIterator<T1>, T3: std::iter::FromIterator<T2>>(
) -> T3
where
    <T1 as std::str::FromStr>::Err: std::fmt::Debug,
{
    std::io::read_to_string(std::io::stdin())
        .expect("PARSE_ERROR")
        .trim()
        .split('\n')
        .map(parse::<T1, T2>)
        .collect::<T3>()
}

fn build_tree(nodes: usize, root: usize, graph: Vec<Vec<usize>>) -> Vec<Vec<usize>> {
    let mut tree = vec![vec![]; nodes + 1];
    let mut stack = vec![root];
    let mut visited = vec![false; nodes + 1];
    visited[root] = true;

    while let Some(node) = stack.pop() {
        for &next in &graph[node] {
            if !visited[next] {
                visited[next] = true;
                tree[node].push(next);
                stack.push(next);
            }
        }
    }
    tree
}

fn max(tree: &Vec<Vec<usize>>, node: usize, values: &mut Vec<Option<usize>>) -> usize {
    let max = tree[node]
        .iter()
        .map(|&node| {
            if let Some(val) = values[node] {
                val
            } else {
                min(tree, node, values)
            }
        })
        .max()
        .unwrap_or(0);

    values[node] = Some(max);
    max
}

fn min(tree: &Vec<Vec<usize>>, node: usize, values: &mut Vec<Option<usize>>) -> usize {
    let min = tree[node]
        .iter()
        .map(|&node| {
            if let Some(val) = values[node] {
                val
            } else {
                max(tree, node, values)
            }
        })
        .min()
        .unwrap_or(0);

    values[node] = Some(min);
    min
}

fn main() {
    let mut input = input::<usize, Vec<_>, std::collections::VecDeque<_>>();

    let line = input.pop_front().unwrap();

    let (nodes, root) = (line[0], line[1]);
    let mut graph = vec![vec![]; nodes + 1];

    (0..(nodes - 1)).for_each(|_| {
        let line = input.pop_front().unwrap();
        let (u, v) = (line[0], line[1]);
        graph[u].push(v);
        graph[v].push(u);
    });

    let line = input.pop_front().unwrap();
    let l = line[0];

    let mut values = vec![None; nodes + 1];

    (0..l).for_each(|_| {
        let line = input.pop_front().unwrap();
        let (node, val) = (line[0], line[1]);
        values[node] = Some(val);
    });

    let tree = build_tree(nodes, root, graph);
    max(&tree, root, &mut values);

    let line = input.pop_front().unwrap();
    let q = line[0];

    let ret = (0..q)
        .map(|_| {
            let line = input.pop_front().unwrap();

            values[line[0]].unwrap().to_string()
        })
        .fold(String::new(), |mut acc, b| {
            acc.push_str(&b);
            acc.push('\n');

            acc
        });

    print!("{}", ret);
}
